Native transfers and ez stats models

mrstepanovic opened the pull request:
This PR introduces:
- 2 new curated models, docs, and tests for native Aleo transfers (silver and gold)
- 2 new stats models, docs, and tests (silver and gold)
- 1 noncore GHA workflow
- 1 updated docfile
- 1 updated test alerts script (`./python/dbt_test_alert.py`) that Jack uses but could switch to datadog

## Why curation was needed for stats
While building the stats models I needed transfers data... these would produce the `unique_from_count` and `unique_to_count` metrics. Aleoscan displays transfer info (I think we chatted about this in Kansas City), so I hoped for a new API endpoint:

![image](https://github.com/user-attachments/assets/686d1b60-3315-4cd7-84f4-27ff3ddc382c)

While there is no such endpoint, Aleoscan is open source and maintains its own backend DB. I found the corresponding [source code](https://github.com/HarukaMa/aleo-explorer/blob/4570ac026d1fec19030f1421a5d9b888029910fd/db/insert.py#L161) with transfer logic, which shows that Aleoscan transfers are parsed through the `credits.aleo` program:

```python
for transition in transitions:
    if transition.program_id == "credits.aleo":
        transfer_from = None
        transfer_to = None
        fee_from = None
        if str(transition.function_name) in ("transfer_public", "transfer_public_as_signer"):
            output = cast(FutureTransitionOutput, transition.outputs[0])
            future = cast(Future, output.future.value)
            transfer_from = str(DatabaseUtil.get_primitive_from_argument_unchecked(future.arguments[0]))
            transfer_to = str(DatabaseUtil.get_primitive_from_argument_unchecked(future.arguments[1]))
            amount = int(cast(u64, DatabaseUtil.get_primitive_from_argument_unchecked(future.arguments[2])))
        elif transition.function_name == "transfer_private_to_public":
            output = cast(FutureTransitionOutput, transition.outputs[1])
            future = cast(Future, output.future.value)
            transfer_to = str(DatabaseUtil.get_primitive_from_argument_unchecked(future.arguments[0]))
            amount = int(cast(u64, DatabaseUtil.get_primitive_from_argument_unchecked(future.arguments[1])))
        elif transition.function_name == "transfer_public_to_private":
            output = cast(FutureTransitionOutput, transition.outputs[1])
            future = cast(Future, output.future.value)
            transfer_from = str(DatabaseUtil.get_primitive_from_argument_unchecked(future.arguments[0]))
            amount = int(cast(u64, DatabaseUtil.get_primitive_from_argument_unchecked(future.arguments[1])))

...
```
 
The `future.arguments[x]` argument mapped to what's stored in the `OUTPUTS` column in `silver__transitions.sql`. But outputs aren't easy to extract: they're typed as strings containing **untyped** lazy JSON:
```json
"id": "7637462966247021801189592432668377709558076887603208062783993621527302681895field",
"type": "future",
"value": "{\n  program_id: credits.aleo,\n  function_name: transfer_public,\n  arguments: [\n    aleo14r3j2sak5n8dwy605j4ada040jajl9ycnxfzesly0urj9azg8sgs0ud6cl,\n    aleo17hwmqhh8annpgm3vqtpd336ww0ywjdeftlw2g3shl6xyt4x3jvxqfwzxhg,\n    1972629u64\n  ]\n}"
```

Using regex in absence of `PARSE_JSON()`, I extracted the arguments as an array whose order matched Aleoscan's. I'm comfortable trusting a scanner as the source of truth... but to [verify](https://aleoscan.io/transaction?id=at1ek6krcuwkcl7g3fnsary60vfz56nazfc0kwspx8k4ph3awzmlvps35vamw):

```sql
select tx_id, transition_id, sender, receiver, amount
from aleo_dev.core.fact_transfers 
where tx_id = 'at1ek6krcuwkcl7g3fnsary60vfz56nazfc0kwspx8k4ph3awzmlvps35vamw';
```

A few other notes:
- Data for private transfers returns null where appropriate.
- All the models in this PR are downstream of `silver__native_transfers.sql`.

## What's next
- Once this PR is merged, deploy new models to the frontend and crosschain repos
- Using the Aleoscan source, add more curated models for validator / staking actions
- Move ahead to [AN-5381](https://team-1612274056224.atlassian.net/browse/AN-5381), observability models


[AN-5381]: https://team-1612274056224.atlassian.net/browse/AN-5381?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

eric-laurello commented on models/silver/core/silver__native_transfers.sql:
```diff
@@ -0,0 +1,141 @@
+    unique_key = ['transition_id'],
+    incremental_strategy = 'merge',
+    merge_exclude_columns = ['inserted_timestamp'],
+    cluster_by = ['block_timestamp::DATE'],
```
Should be modified_timestamp for the cluster in silver since that's how we query it in the gold model

eric-laurello commented on models/silver/core/silver__native_transfers.sql:
```diff
@@ -0,0 +1,141 @@
+            10,
+            6
+        ) AS amount,
+    'aleo_credits' as currency,
```
should think about this name - not sure what Aleo calls it, but for other chains we typically have gone with token_address

eric-laurello commented on models/silver/core/silver__native_transfers.sql:
```diff
@@ -0,0 +1,141 @@
+            6
+        ) AS amount,
+    'aleo_credits' as currency,
+    inserted_timestamp,
```
add the sysdate and the actual invocation id here for this run

eric-laurello commented on models/silver/core/silver__native_transfers.sql:
```diff
@@ -0,0 +1,141 @@
+    WHERE
+        program_id = 'credits.aleo'
+        AND function IN (
+            'transfer_public',
```
I looked briefly and a couple of the other functions. Maybe we need to capture the bond/unbound as well since credits are technically leaving the account.
[https://aleoscan.io/transition?id=au1xwc4dqu9fmyh6ypf2r40zgz4nxej0tw85ykzvlv7j7q4kzg0c5pqk5n8kt](url)

mrstepanovic replied:
I thought a lot about this and would prefer to break out bond/unbond events into a separate model. Most of the commentary and tutorials I've seen have a very limited set of functions considered transfers. Like this: https://medium.com/@atletiko003/transferring-aleo-credits-a-comprehensive-tutorial-e253ad48f8bc

eric-laurello replied:
yeah it makes sense to not overcomplicate this model. Was thinking more about users trying to get to a balance from transfers, but that will inherently be impossible on aleo.

eric-laurello commented on models/silver/stats/silver_stats__core_metrics_hourly.sql:
```diff
@@ -0,0 +1,82 @@
+            WHEN NOT succeeded THEN tx_id
+        END
+    ) AS transaction_count_failed,
+    COUNT(
```
Two Ideas here - just got this idea after seeing this.

1. Add fee payer to the transactions table
2. Use fee payer as the unique count
  a. current method is only counting for tx that have >1 transfer
  
  How we might get fee payer:
  `    select   outputs,    REGEXP_SUBSTR(
            outputs[array_size(outputs)-1] :value :: STRING,
            'arguments:\\s*\\[(.*?)\\]',
            1,
            1,
            'sie'
        ) as args_string,
           SPLIT(
            REGEXP_REPLACE(
                REGEXP_REPLACE(
                    REGEXP_REPLACE(args_string, '\\s+', ''),
                    '\\[|\\]',
                    ''
                ),
                'u64$',
                ''
            ),
            ','
        ) aa, aa[0]::STRING as fee_payer from aleo_dev.silver.transitions_fee where tx_id = 'at1jdghl9umq7dvzd4z548x0hv6gsu5pfy7320xam38flnzqsx5xczqfyl0qq'`

mrstepanovic replied:
If we switch to fee payers for `unique_from_count` I'm not sure what to do about `unique_to_count`. We'd have 1 fee payer per transaction and can count distinct there. But I think we'd need to know how to parse all program functions before we can get `unique_to_count` this way.

eric-laurello replied:
we can kill the to count. Many of the chains only have one value (sender, signer etc) -- check aptos or solana for example

eric-laurello commented on models/silver/stats/silver_stats__core_metrics_hourly.sql:
```diff
@@ -0,0 +1,82 @@
+    '{{ invocation_id }}' AS _invocation_id
+
+FROM
+    {{ ref('silver__transitions') }} ts
```
use the gold table, that way we dont need to join back to the transition table to get the fee

mrstepanovic commented:
Addressed most of your comments, plus:

### Added observability models
- silver_observability__blocks_completeness
- silver_observability__transactions_completeness
- silver_observability__transitions_completeness
- (Still need to add third-party LQ or second SL wellness check)
## Modified transfer handling
- Replaced currency with token_address and is_native flags (I saw some other chains do this. Can later union in `is_native = FALSE` addresses when needed)
- Updated fee tracking to capture fee_payer from transition outputs
- Simplified unique address counting to use fee payer instead of sender/receiver
## Fixed a few tests
- Corrected test_core__transfers_recent to test_core__transfers_full
- Added a few core model recency tests
## Other things and TODOs:
- Moved program_id extraction to transaction level
- Need to add a bond/unbond table, though that is more challenging as bond events and unbond events have different data structures

eric-laurello commented on models/gold/stats/stats__ez_core_metrics_hourly.sql:
```diff
@@ -0,0 +1,25 @@
+    inserted_timestamp,
+    modified_timestamp
+FROM
+    {{ ref('silver_stats__core_metrics_hourly') }}
```
add incremental logic

eric-laurello commented on models/silver/stats/silver_stats__core_metrics_hourly.sql:
```diff
@@ -0,0 +1,76 @@
+
+FROM
+    {{ ref('silver__transitions_fee') }} ts
+JOIN
```
can kill this join too?

mrstepanovic replied:
I added in-line github comments into cursor and it's the worst thing ever. meant to do this anyway

eric-laurello commented on models/silver/core/silver__transitions_fee.sql:
```diff
@@ -68,6 +68,18 @@ fee_sum AS (
     GROUP BY
         transition_id
+),
+fee_payer AS (
```
can we add the fee payer to the gold fact_transactions table?

mrstepanovic commented on models/silver/stats/silver_stats__core_metrics_hourly.sql:
```diff
@@ -0,0 +1,72 @@
+    SYSDATE() AS modified_timestamp,
+    '{{ invocation_id }}' AS _invocation_id
+FROM
+    {{ ref('silver__transitions_fee') }} ts
```
While we can get all the data from this model, I don't think it's sufficient for edge cases. like if there are no transactions / transitions in a block, it might introduce a block gap. I might want to switch this to gold, too

mrstepanovic replied:
Like, Lava has 2 silver stats models for blocks and metrics then joins them in gold

eric-laurello replied:
is it possible to have 0 tx blocks?

eric-laurello replied:
yeah, good catch:
`select block_id from aleo.core.fact_blocks
      except 
      
      select block_id from aleo.core.fact_transactions`

1M blocks with txs

eric-laurello commented on models/silver/core/silver__native_transfers.sql:
```diff
@@ -0,0 +1,139 @@
+    unique_key = ['transition_id'],
+    incremental_strategy = 'merge',
+    merge_exclude_columns = ['inserted_timestamp'],
+    cluster_by = ['modified_timestamp::DATE'],
```
Can you add block_timestamp::Date here too. I know we arent making this user facing, but the inc predicate will use it

eric-laurello commented on models/silver/core/silver__native_transfers.yml:
```diff
@@ -0,0 +1,92 @@
+    tests:
+      - dbt_utils.unique_combination_of_columns:
+          combination_of_columns:
+            - TX_ID
```
i think we can just do transition_id

eric-laurello commented on models/gold/core/core__fact_transactions.sql:
```diff
@@ -79,6 +79,7 @@ SELECT
         fee,
         0
     ) AS fee,
+    b.fee_payer,
```
Add SO for fee_payer too

eric-laurello replied:
also add the column to the yml doc

eric-laurello commented on models/gold/core/core__fact_transfers.sql:
```diff
@@ -0,0 +1,47 @@
+{{ config(
```
update the overview doc to add this model

eric-laurello commented on models/silver/stats/silver_stats__core_metrics_hourly.sql:
```diff
@@ -0,0 +1,66 @@
+        'hour',
+        CURRENT_TIMESTAMP
+    )
+{% if is_incremental() %}
```
can kill this as it wont work in the view

eric-laurello commented on models/silver/stats/silver_stats__core_metrics_hourly.sql:
```diff
@@ -0,0 +1,66 @@
+
+{{ config(
+    materialized = 'view',
+    tags = ['noncore','recent_test']
```
remove recent_test tag, dont want to test the view

eric-laurello commented on models/silver/stats/silver_stats__core_metrics_block_hourly.sql:
```diff
@@ -0,0 +1,27 @@
+{{ config(
+    materialized = 'view',
+    tags = ['noncore','recent_test']
```
remove recent_test tag, dont want to test the view
